design/ip-address-management-for-networkdata.md

```diff
@@ -0,0 +1,328 @@
+## Design Details
+
+
+A Metal3IPPool object would be created :
+
+```yaml
+apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
+kind: Metal3IPPool
+metadata:
+  name: pool-1
+  namespace: default
+  ownerReferences:
+  - apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
+    kind: Metal3Cluster
+    name: cluster-1
+  - apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
+    kind: Metal3Data
+    name: metal3data-1
+spec:
+  pools:
+    - start: 192.168.0.10
+      end: 192.168.0.15
+      subnet: 192.168.0.0/24
+      step: 1
+    - start: 192.168.0.20
+      end: 192.168.0.25
+      subnet: 192.168.0.0/24
+status:
+  lastUpdated: "2020-04-02T06:36:09Z"
+  allocations:
+    "192.168.0.11": "metal3data-1"
+  addresses:
+    "metal3data-1": "pool-1-192-168-0-11"
+  error: false
+  errorMessage: ""
+```
+
+In the *spec* of *Metal3IPPool*, there would be a *pools* list, that would
+contain a list of IP address pools, with the *start* and *end* attributes
+giving the start and end ip addresses of the pool, the *subnet* allowing to
+verify that the allocated IP is in the pool and from which the start and end ip
+addresses can be inferred. The *step* item gives the interval between allocated
+IP addresses. Specifying single ip addresses can be achieved by setting the
+start and end ip address to that single ip address.
+
+The *status* would contain a *lastUpdated* field with the timestamp of the last
+update and and *error* boolean that is set to true if an error during allocation
+happens. In that case, the *errorMessage* will contain the text of the error.
+The *allocations* map will map the IP address to the *Metal3Data* object it was
+allocated for and the *addresses* will map the *metal3Data* objects with the
+*Metal3IPAddress* objects.
+
+All IP addresses specified in a Metal3IPPool must be from the same subnet.
+
+The *Metal3IPAddress* object would be the following
+
+```yaml
+apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
+kind: Metal3IPAddress
+metadata:
+  name: pool-1-192-168-0-11
+  namespace: default
+  ownerReferences:
+  - apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
+    kind: Metal3IPPool
+    name: pool-1
+  - apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
+    kind: Metal3Data
+    name: metal3data-1
+spec:
+  Metal3Data:
+    Name: metal3data-1
+  Address: 192.168.0.11
+  Metal3IPPool:
+    Name: pool-1
+status:
+  ready: true
+```
+
+For each owner reference added on a *Metal3IPPool* that is a *Metal3Data*, the
+controller reconciling the *Metal3IPPool* will select an available IP address.
+If all the *Metal3Data* owner references are from the same *Metal3DataTemplate*,
+then the controller will use the index of the *Metal3Data* to compute the IP
+address to allocate by sorting the addresses in incremental order and selecting
+the IP address corresponding to the index. In case of conflict, (multiple
+Metal3DataTemplate referencing the same *Metal3IPPool*, a random available IP
+address will be selected.
```

Comment:
This algorithm sounds more complicated than necessary. Why not just pick an IP randomly from the pool? Or sequentially based on the order of the owner references?

Why make the pool controller aware of the Metal3Data type? Why does the pool care who is asking for an IP?

Reply:
The idea behind the algorithm was to keep the feature parity with the previous proposal, where the ip address is tied to the index. this gives a predictable allocation, that is of great interest if you are using something that would benefit from IP addresses being across upgrade. It allows you to know the ip addresses of all the node in a nodepool predictably, for example if you are using Metallb and need to configure the BGP speakers on your ToR router. Having predictable and stable IP addresses in that case is necessary.

Reply:
I had been assuming the controller for Metal3IPPool was independent of the Metal3DataTemplate controller, but based on your comment here it sounds like it isn't and that the "index" is coming from the template. Is that right?

I feel like we're trying to force predictability in a system designed not to worry about the order things happen, and I'm worried that's going to give us something more complex as a result. If we need to know the IPs of hosts, it seems like those should be set explicitly and declaratively, rather than asking something to pick an IP from a pool.

Reply:
We could actually add a feature regarding predictability based on the object names. like a map of name and reserved IP addresses. then we would remove the requirement based on the index. I'll give it a shot in the next commit